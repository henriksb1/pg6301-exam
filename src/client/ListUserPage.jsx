import * as React from "react";
import { ErrorView } from "./ErrorView";
import { LoadingView } from "./LoadingView";
import { useLoading } from "./useLoading";
import { Link } from "react-router-dom";

export function ListUserPage({ userApi }) {
  const { data: user, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !user) {
    return <LoadingView />;
  }

  return (
    <div>
      <h1>Edit users</h1>
      {user.map(({ id, firstName, lastName, email }) => (
        <li key={id}>
          <Link to={`/user/${id}/edit`}>
            {firstName + " " + lastName + ` (${email})`}
          </Link>
        </li>
      ))}
    </div>
  );
}
