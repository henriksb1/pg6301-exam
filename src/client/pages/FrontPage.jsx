import React from "react";

export function FrontPage() {
  return (
    <div>
      <h1>Welcome 👋🏻</h1>
      <p>Please log in to use this messaging application</p>
    </div>
  );
}
