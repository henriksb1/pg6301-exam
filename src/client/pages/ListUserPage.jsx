import * as React from "react";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";

export function ListUserPage({ userApi }) {
  const { data: user, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !user) {
    return <LoadingView />;
  }

  return (
    <div>
      <h1>All users</h1>
      {user.map(({ id, firstName, lastName, email }) => (
        <li key={id}>{firstName + " " + lastName + ` (${email})`}</li>
      ))}
    </div>
  );
}
