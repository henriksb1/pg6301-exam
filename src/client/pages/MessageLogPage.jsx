import React from "react";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";
import { fetchJson } from "../lib/http";

export function MessageLogPage({ messageApi }) {
  const { data: user, error, loading, reload } = useLoading(
    async () => await messageApi.listMessages()
  );

  const {
    data: userData,
    error: errorData,
    loading: loadingData,
  } = useLoading(() => fetchJson("/api/profile"));

  if (error || errorData) {
    return <ErrorView error={error ? error : errorData} reload={reload} />;
  }

  if (loading || !user || loadingData) {
    return <LoadingView />;
  }

  const { username } = userData;
  const findSentMessages = user.filter((e) => e.username === username);
  console.log(findSentMessages);
  const findReceivedMessages = user.filter((e) => e.receiver === username);

  return (
    <div>
      <h1>Message log</h1>
      <h2>Sent messages</h2>
      {findSentMessages.map(({ id, username, message }) => (
        <div className={"chat"} key={id}>
          <div className={"chat-content"}>
            <p>{"Sender: " + username}</p>
            <p>{"Message: " + message}</p>
          </div>
        </div>
      ))}
      <h2>Received Messages </h2>
      {findReceivedMessages.map(({ id, username, message }) => (
        <div className={"chat"} key={id}>
          <p>{"Sender: " + username}</p>
          <p>{"Message: " + message}</p>
        </div>
      ))}
    </div>
  );
}
