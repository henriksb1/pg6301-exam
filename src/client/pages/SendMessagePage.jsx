import React, { useState } from "react";
import { InputField } from "../components/InputField";
import { useLoading } from "../lib/useLoading";
import { fetchJson } from "../lib/http";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";

export function SendMessagePage({ userApi, messageApi }) {
  const [receiver, setReceiver] = useState("");
  const [message, setMessage] = useState("");

  const { data: profileData, profileError, profileLoading } = useLoading(() =>
    fetchJson("/api/profile")
  );

  const { data: user, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error || profileError) {
    return <ErrorView error={error ? error : profileError} />;
  }

  if (profileLoading || !user) {
    return <LoadingView />;
  }

  const { username } = profileData;

  async function submit(e) {
    e.preventDefault();
    await messageApi.createMessage({ username, receiver, message });
  }

  return (
    <form onSubmit={submit}>
      <h1>Send new message</h1>
      <InputField
        label={"Message"}
        value={message}
        onValueChange={setMessage}
      />
      <label>
        Receiver:
        <select onChange={(e) => setReceiver(e.target.value)}>
          <option />
          {user.map(({ id, email }) => (
            <option key={id}>{email}</option>
          ))}
        </select>
      </label>
      <br />
      <button>Send</button>
    </form>
  );
}
