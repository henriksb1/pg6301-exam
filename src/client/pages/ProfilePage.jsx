import React from "react";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { fetchJson } from "../lib/http";
import { RegisterUserPage } from "./RegisterUserPage";
import { ListUserPage } from "./ListUserPage";

export function ProfilePage({ userApi }) {
  const { loading, error, data } = useLoading(() => fetchJson("/api/profile"));

  if (error) {
    return <ErrorView error={error} />;
  }
  if (loading || !data) {
    return <LoadingView />;
  }

  const { username } = data;

  return (
    <div>
      <h1>Your profile:</h1>
      <div>Email: {username}</div>
      <RegisterUserPage userApi={userApi} />
      <ListUserPage userApi={userApi} />
    </div>
  );
}
