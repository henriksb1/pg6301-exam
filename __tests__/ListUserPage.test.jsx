import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import { ListUserPage } from "../src/client/pages/ListUserPage";

const userApi = {
  listUsers: async () => [
    {
      id: 1,
      firstName: "Per",
      lastName: "Borgli",
      email: "per@borgli",
      password: "123",
    },
  ],
};

describe("list users component", () => {
  it("should show users on dom", async () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    await act(async () => {
      ReactDOM.render(
        <MemoryRouter>
          <ListUserPage userApi={userApi} />
        </MemoryRouter>,
        container
      );
    });

    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("li").textContent).toEqual(
      "Per Borgli (per@borgli)"
    );
  });
});
