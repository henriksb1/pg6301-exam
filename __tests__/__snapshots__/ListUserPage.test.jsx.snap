// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list users component should show error message 1`] = `"<div>An error occurred: Error: Failed to load</div>"`;

exports[`list users component should show loading screen 1`] = `"<div>Loading...</div>"`;

exports[`list users component should show users on dom 1`] = `"<div><h1>All users</h1><li>Per Borgli (per@borgli)</li></div>"`;
